[Theme] add alpha override support to references and hex colors.

auregosu opened the pull request:
this adds support for the syntax `@color / 50%;` and `#123456 / 50%` to override the alpha of a given color.
- extended grammar rules for `t_property_element` and `t_property_color`
- added `gboolean` member to link `union` to differentiate modified links from normal links
- added definition of `rofi_theme_resolve_modified_link()`
- added implementation of modified link resolving in `rofi_theme_parse_process_links_int()`

https://github.com/davatorium/rofi/discussions/2200

auregosu committed:
[Theme] add alpha override support to references and hex colors.
this adds support for the syntax `@color / 50%;` and `#123456 / 50%` to
override the alpha of a given color.
- extended grammar rules for `t_property_element` and `t_property_color`
- added `gboolean` member to link `union` to differentiate modified
  links from normal links
- added definition of `rofi_theme_resolve_modified_link()`
- added implementation of modified link resolving in
  `rofi_theme_parse_process_links_int()`

DaveDavenport commented on lexer/theme-parser.y:
```diff
@@ -999,8 +1006,9 @@ t_property_color
-| T_COLOR {
+| T_COLOR t_property_color_opt_alpha_ws {
     $$ = $1;
+    $$.alpha  = $2;
```
This overrides the alpha that is (possibly) already set.

maybe a `*=`  instead of `=`?

auregosu replied:
i made it analogous to the already existing behavior for named colors on line 1019:
```
| T_COLOR_NAME t_property_color_opt_alpha_ws {
    $$ = $1;
    $$.alpha = $2;
}
```
it's probably fine to leave it this way.

DaveDavenport commented on source/theme.c:
```diff
@@ -1592,6 +1592,16 @@ static void rofi_theme_parse_process_links_int(ThemeWidget *wid) {
+        if (pv->value.link.ref == NULL) {
+			rofi_theme_resolve_modified_link(pv);
+			if (pv->value.link.ref) {
+				ThemeColor referenced_color = pv->value.link.ref->value.color;
```
is it guaranteed that the 'link'  points to a color?

auregosu replied:
i'll add a check for the type of `pv->value.link.ref` on line 1598. should there be a warning if it's otherwise?

DaveDavenport replied:
A g_debug should be fine and the color should not be set. 
(I think this is normal behaviour if you try to set color to f.e. a boolean..   the color is not set, and 'Theme entry: ... unset"  is printed in debug.)

auregosu commented:
sorry for force pushing yet again. i corrected the alpha overriding of hex colors so it only happens when `t_property_color_opt_alpha_ws` is set. i also fixed a naive mix-up of string assignment and copying that lead to other errors.

auregosu committed:
Merge branch 'next' into next
